src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -598,6 +602,32 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
             }
         }
 
+        private void VerifyAndUpdateWrite(int count)
+        {
+            TryInitializeResponseContentLength();
+
+            if (_responseContentLength.HasValue && _responseBytesWritten + count > _responseContentLength.Value)
+            {
+                _keepAlive = false;
+                throw new InvalidOperationException(
+                    $"Response Content-Length mismatch: too many bytes written ({_responseBytesWritten + count} of {_responseContentLength.Value}).");
+            }
+
+            _responseBytesWritten += count;
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
+            {
+                var responseHeaders = FrameResponseHeaders;
+                if (responseHeaders != null && responseHeaders.HasContentLength)
+                {
+                    _responseContentLength = long.Parse(responseHeaders.HeaderContentLength.ToString(), NumberStyles.None);
```

Comment:
This will result in a FormatException being thrown from Write. Use TryParse and ignore values you can't parse.
When determining if you should chunk, do you check for the presence of a content-length or if you understand it?